Improve compression ratio of levels 3 & 4

Cyan4973 opened the pull request:
Following #4170 , there were a few ideas I wanted to test for the general compression performance of level 3, aka `dfast` strategy.

The one proposed in this PR is consistently positive.
It feels retrospectively obvious: select the supplemental "long" match candidate only if it's effectively longer.
On `silesia.tar` corpus, this modification saves ~75 KB at level 3. 

And thankfully, the measured speed cost is negligible, below noise level (between 0 and -1% depending on measurements).

Cyan4973 committed:
slightly improved compression ratio at levels 3 & 4

The compression ratio benefits are small but consistent, i.e. always positive.
On `silesia.tar` corpus, this modification saves ~75 KB at level 3.
The measured speed cost is negligible, i.e. below noise level, between 0 and -1%.

Cyan4973 committed:
fix test

a margin of 4 is insufficient to guarantee compression success.

terrelln commented on lib/compress/zstd_double_fast.c:
```diff
@@ -253,19 +253,23 @@ size_t ZSTD_compressBlock_doubleFast_noDict_generic(
+                if (llen > mLength) {
+                    ip = ip1;
+                    mLength = llen;
+                    offset = (U32)(ip-matchl1);
+                    while (((ip>anchor) & (matchl1>prefixLowest)) && (ip[-1] == matchl1[-1])) { ip--; matchl1--; mLength++; } /* catch up */
+                    goto _match_found;
```
Could this code be merged with the code in lines 273-275?

We could set `matchs0 = matchl1`, and then just fallthrough to the code on line 273. Though we might want to rename some variables.

Cyan4973 committed:
rewrite code in the manner suggested by @terrelln

Cyan4973 committed:
enable regression tests at pull request time

was transferred from circleci,
but was only triggered on push into dev,
i.e. after pull request is merged.

Cyan4973 committed:
updated regression tests results
